consensus/istanbul/core/vrank.go

```diff
@@ -0,0 +1,240 @@
+
+	committedTime := v.TimeSinceStart()
+	if v.threshold > committedTime {
+		v.threshold = committedTime
+	}
+
+	vrankQuorumCommitArrivalTimeGauge.Update(int64(committedTime))
+	if len(v.commitArrivalTimeMap) != 0 {
+		sum := int64(0)
+		for _, v := range v.commitArrivalTimeMap {
+			sum += int64(v)
+		}
+		avg := sum / int64(len(v.commitArrivalTimeMap))
+		vrankAvgCommitArrivalTimeWithinQuorumGauge.Update(avg)
+	}
+}
+
+// Log logs accumulated data in a compressed form
+func (v *Vrank) Log() (string, []string) {
+	var (
+		serialized = serialize(v.committee, v.commitArrivalTimeMap)
+		assessed   = assessBatch(serialized, v.threshold)
+		compressed = compress(assessed)
+		bitmap     = hex.EncodeToString(compressed)
+
+		lateCommits = make([]time.Duration, 0)
+		lastCommit  = time.Duration(0)
+	)
+
+	for _, t := range serialized {
+		if assess(t, v.threshold) == vrankArrivedLate {
+			lateCommits = append(lateCommits, t)
+		}
+	}
+
+	// lastCommit = max(lateCommits)
+	for _, t := range lateCommits {
+		if lastCommit < t {
+			lastCommit = t
+		}
+	}
+	if lastCommit != time.Duration(0) {
+		vrankLastCommitArrivalTimeGauge.Update(int64(lastCommit))
+	}
+
+	lateCommitsStrArr := encodeDurationBatch(lateCommits)
+
+	logger.Info("VRank", "seq", v.view.Sequence.Int64(),
+		"round", v.view.Round.Int64(),
+		"bitmap", bitmap,
+		"late", lateCommitsStrArr,
+	)
+	return bitmap, lateCommitsStrArr
+}
+
+func (v *Vrank) isTargetCommit(msg *istanbul.Subject, src istanbul.Validator) bool {
+	if msg.View.Cmp(&v.view) != 0 {
+		return false
+	}
+	_, ok := v.commitArrivalTimeMap[src.Address()]
+	if ok {
+		return false
+	}
+	return true
+}
+
+// assess determines if given time is early, late, or not arrived
+func assess(t, threshold time.Duration) int {
+	if t == vrankNotArrivedPlaceholder {
+		return vrankNotArrived
+	}
+
+	if t > threshold {
+		return vrankArrivedLate
+	} else {
+		return vrankArrivedEarly
+	}
+}
+
+func assessBatch(ts []time.Duration, threshold time.Duration) []int {
+	ret := make([]int, 0)
+	for _, t := range ts {
+		ret = append(ret, assess(t, threshold))
+	}
+	return ret
+}
+
+// serialize serializes arrivalTime hashmap into array.
+// If committee is sorted, we can simply figure out the validator position in the output array
+// by sorting the output of `klay.getCommittee()`
+func serialize(committee istanbul.Validators, arrivalTimeMap map[common.Address]time.Duration) []time.Duration {
+	sortedCommittee := make(istanbul.Validators, len(committee))
+	copy(sortedCommittee[:], committee[:])
+	sort.Sort(sortedCommittee)
+
+	serialized := make([]time.Duration, len(sortedCommittee))
+	for i, v := range sortedCommittee {
+		val, ok := arrivalTimeMap[v.Address()]
+		if ok {
+			serialized[i] = val
+		} else {
+			serialized[i] = vrankNotArrivedPlaceholder
+		}
+
+	}
+	return serialized
+}
+
+// compress compresses data into 2-bit bitmap
+// e.g., [1, 0, 2] => [0b01_00_10_00]
+func compress(arr []int) []byte {
+	zip := func(a, b, c, d int) byte {
```

Comment:
I didn't think any code will use this function. I'm not sure even if this will be used in the future other than logging.

Reply:
Because I was concerned if it consumes unnecessary overhead, compared to a case that defines to the outer function. If the function does not capture anything, however, I guess it does not may create another overhead if the compiler is smart by analyzing that there's no capturing, so it may decide that it is safe to move the function to the outer scope.

Anyway, my context was not visibility, but the performance aspect.